presto-main/src/main/java/com/facebook/presto/operator/StreamingAggregationOperator.java

```diff
@@ -0,0 +1,294 @@
+    }
+
+    @Override
+    public boolean needsInput()
+    {
+        return !finishing && pendingInput == null;
+    }
+
+    @Override
+    public void addInput(Page page)
+    {
+        checkState(pendingInput == null);
+        requireNonNull(page, "page is null");
+
+        processInput(page);
+        updateMemoryUsage();
+    }
+
+    private void updateMemoryUsage()
+    {
+        long memorySize = pageBuilder.getRetainedSizeInBytes();
+        for (Aggregator aggregator : aggregates) {
+            memorySize += aggregator.getEstimatedSize();
+        }
+
+        if (previousRow != null) {
+            memorySize += previousRow.getRetainedSizeInBytes();
+        }
+
+        if (step.isOutputPartial()) {
+            systemMemoryContext.setBytes(memorySize);
+        }
+        else {
+            userMemoryContext.setBytes(memorySize);
+        }
+    }
+
+    private void processInput(Page page)
+    {
+        requireNonNull(page, "page is null");
+
+        if (previousRow != null) {
+            if (!rowEqualsRow(0, previousRow, 0, page)) {
+                evaluateAggregatesToPageBuilder(previousRow, 0);
+                previousRow = null;
+
+                if (pageBuilder.isFull()) {
+                    pendingInput = page;
+                    return;
+                }
+            }
+            previousRow = null;
+        }
+
+        int startPosition = 0;
+        while (true) {
+            int groupEnd = findGroupEnd(startPosition, page);
+            addRowsToAggregates(page, startPosition, groupEnd);
+
+            if (groupEnd < page.getPositionCount()) {
+                evaluateAggregatesToPageBuilder(page, startPosition);
+
+                if (pageBuilder.isFull()) {
+                    pendingInput = page.getRegion(groupEnd, page.getPositionCount() - groupEnd);
+                    return;
+                }
+
+                startPosition = groupEnd;
+            }
+            else {
+                previousRow = page.getRegion(page.getPositionCount() - 1, 1);
+                pendingInput = null;
+                return;
+            }
+        }
+    }
+
+    private void addRowsToAggregates(Page page, int startPosition, int endPosition)
+    {
+        for (Aggregator aggregator : aggregates) {
+            aggregator.processPage(page.getRegion(startPosition, endPosition - startPosition));
+        }
+    }
+
+    private void evaluateAggregatesToPageBuilder(Page page, int position)
+    {
+        pageBuilder.declarePosition();
+        for (int i = 0; i < groupByTypes.size(); i++) {
+            Block block = page.getBlock(groupByChannels.get(i));
+            Type type = groupByTypes.get(i);
+            type.appendTo(block, position, pageBuilder.getBlockBuilder(i));
+        }
+        int offset = groupByTypes.size();
+        for (int i = 0; i < aggregates.size(); i++) {
+            aggregates.get(i).evaluate(pageBuilder.getBlockBuilder(offset + i));
+        }
```

Comment:
I would add builder full check here and add produced pages to some `outputPages` list. Then you wouldn't have to worry about `pendingInput`.

Reply:
@sopel39 Karol, thank you for being so patient with me. I tried this suggestion and it did simplify the code. Thanks!